File: app/api/inventory.py
"""库存路由"""
from typing import List
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session, joinedload
from app.database import get_db
from app.models import InventoryItem, InventoryStatus, ItemDefinition
from app.schemas import CreateInventoryItemRequest, InventoryItemOut
from app.deps import get_current_user_id
from app.errors import NotFound, BadRequest, Forbidden

router = APIRouter(prefix="/api/inventory", tags=["库存"])


@router.get("", response_model=List[InventoryItemOut])
def list_inventory(
    status: str = None,
    category: str = None,
    search: str = None,
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """
    获取当前用户的库存列表
    
    筛选条件:
    - status: 按库存状态筛选 (available/locked/traded/removed)
    - category: 按饰品品类筛选 (weapon/knife/gloves/sticker/agent)
    - search: 按饰品名称关键词搜索
    
    性能: 使用 joinedload 预加载 definition 关联，避免 N+1 查询
    """
    query = db.query(InventoryItem).options(joinedload(InventoryItem.definition))
    query = query.filter(InventoryItem.owner_id == user_id)

    if status:
        query = query.filter(InventoryItem.status == status)
    if category:
        query = query.join(ItemDefinition).filter(ItemDefinition.category == category)
    if search:
        query = query.join(ItemDefinition).filter(ItemDefinition.name.contains(search))

    items = query.order_by(InventoryItem.created_at.desc()).all()
    return [InventoryItemOut.from_orm(i) for i in items]


@router.post("", response_model=InventoryItemOut, status_code=201)
def create_inventory_item(
    req: CreateInventoryItemRequest,
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """
    添加饰品到当前用户的库存
    
    校验:
    - definition_id 必须存在于 item_definitions 数据字典
    
    创建后:
    - 状态默认为 available
    - 需重新查询以加载 definition 关联关系
    """
    # 验证装备字典存在
    definition = db.query(ItemDefinition).filter(ItemDefinition.id == req.definition_id).first()
    if not definition:
        raise NotFound("装备不存在于数据字典中")

    item = InventoryItem(
        owner_id=user_id,
        definition_id=req.definition_id,
        quality=req.quality,
        float_value=req.float_value,
        pattern=req.pattern,
        stat_trak=req.stat_trak,
        souvenir=req.souvenir,
        description=req.description,
        image_url=req.image_url,
    )
    db.add(item)
    db.flush()
    db.refresh(item)

    # 重新查询以加载关联关系
    item = db.query(InventoryItem).options(
        joinedload(InventoryItem.definition)
    ).filter(InventoryItem.id == item.id).first()

    return InventoryItemOut.from_orm(item)


@router.delete("/{item_id}")
def delete_inventory_item(
    item_id: int,
    user_id: int = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """
    移除库存中的饰品
    
    约束:
    - 只能移除 available 状态的饰品（上架中或锁定中不可删除）
    - 只能删除自己的饰品
    
    实际操作为软删除: 将状态改为 removed 而非物理删除
    """
    item = db.query(InventoryItem).filter(InventoryItem.id == item_id).first()
    if not item:
        raise NotFound("饰品不存在")
    if item.owner_id != user_id:
        raise Forbidden("无权操作此饰品")
    if item.status != InventoryStatus.available:
        raise BadRequest("饰品当前状态不允许删除")

    item.status = InventoryStatus.removed
    db.flush()

    return {"message": "饰品已移除"}

File: src/routes/inventory.ts
import { Router, Request, Response } from "express";
import { z } from "zod";
import { authMiddleware } from "../middleware/auth";
import { validate } from "../middleware/validate";
import { AppError } from "../middleware/errorHandler";

export const inventoryRouter = Router();

const createItemSchema = z.object({
  definitionId: z.number({ coerce: true }),
  quality: z.enum(["FN", "MW", "FT", "WW", "BS"]),
  floatValue: z.number().min(0).max(1).optional(),
  pattern: z.number().int().optional(),
  statTrak: z.boolean().optional().default(false),
  souvenir: z.boolean().optional().default(false),
  description: z.string().max(500).optional(),
  imageUrl: z.string().url().optional(),
});

// ---- 获取我的库存 ----
inventoryRouter.get("/", authMiddleware, async (req: Request, res: Response) => {
  const { prisma } = await import("../index");
  const { status, category, quality, search } = req.query;

  const where: any = { ownerId: req.user!.userId };
  if (status) where.status = status;
  if (quality) where.quality = quality;
  if (category) {
    where.definition = { category: category as any };
  }
  if (search) {
    where.definition = {
      ...(where.definition || {}),
      name: { contains: search as string },
    };
  }

  const items = await prisma.inventoryItem.findMany({
    where,
    include: { definition: true },
    orderBy: { createdAt: "desc" },
  });

  res.json({ items });
});

// ---- 添加饰品到库存 ----
inventoryRouter.post("/", authMiddleware, validate(createItemSchema), async (req: Request, res: Response) => {
  const { prisma } = await import("../index");
  const data = req.body;

  // 验证 definition 存在
  const def = await prisma.itemDefinition.findUnique({ where: { id: data.definitionId } });
  if (!def) throw new AppError(404, "装备不存在于数据字典中");

  const item = await prisma.inventoryItem.create({
    data: {
      ownerId: req.user!.userId,
      definitionId: data.definitionId,
      quality: data.quality,
      floatValue: data.floatValue,
      pattern: data.pattern,
      statTrak: data.statTrak,
      souvenir: data.souvenir,
      description: data.description,
      imageUrl: data.imageUrl,
    },
    include: { definition: true },
  });

  res.status(201).json({ item });
});

// ---- 删除库存饰品 ----
inventoryRouter.delete("/:id", authMiddleware, async (req: Request, res: Response) => {
  const { prisma } = await import("../index");
  const itemId = BigInt(req.params.id);

  const item = await prisma.inventoryItem.findUnique({ where: { id: itemId } });
  if (!item) throw new AppError(404, "饰品不存在");
  if (item.ownerId !== req.user!.userId) throw new AppError(403, "无权操作此饰品");
  if (item.status !== "available") throw new AppError(400, "饰品当前状态不允许删除");

  await prisma.inventoryItem.update({
    where: { id: itemId },
    data: { status: "removed" },
  });

  res.json({ message: "饰品已移除" });
});
